stan/math/rev/functor/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,248 @@
+ * @return Eigen object with result of applying functor to inputs.
+ */
+template <typename T1, typename T2, typename F,
+          require_std_vector_vt<is_std_vector, T1>* = nullptr,
+          require_std_vector_st<std::is_integral, T1>* = nullptr,
+          require_var_matrix_t<T2>* = nullptr>
+inline auto apply_scalar_binary(const T1& x, const T2& y, const F& f) {
+  return f(x, y);
+}
+
+/**
+ * Specialisation for use when the first input is an `var_value<Eigen> type and
+ * the second is a scalar.
+ *
+ * @tparam T1 Type of `var_value<Matrix>` object to which functor is applied.
+ * @tparam T2 Type of scalar to which functor is applied.
+ * @tparam F Type of functor to apply.
+ * @param x Matrix input to which operation is applied.
+ * @param y Scalar input to which operation is applied.
+ * @param f functor to apply to var matrix and scalar inputs.
+ * @return `var_value<Matrix> object with result of applying functor to inputs.
+ *
+ * Note: The return expresssion needs to be evaluated, otherwise the captured
+ *         function and scalar fall out of scope.
+ */
+template <typename T1, typename T2, typename F,
+          require_var_matrix_t<T1>* = nullptr,
+          require_stan_scalar_t<T2>* = nullptr>
+inline auto apply_scalar_binary(const T1& x, const T2& y, const F& f) {
+  return f(x, y);
+}
+
+/**
+ * Specialisation for use when the first input is an scalar and the second is
+ * an `var_value<Eigen>`.
+ *
+ * @tparam T1 Type of scalar to which functor is applied.
+ * @tparam T2 var value with inner Eigen type to which functor is applied.
+ * @tparam F Type of functor to apply.
+ * @param x Scalar input to which operation is applied.
+ * @param y var matrix input to which operation is applied.
+ * @param f functor to apply to Eigen and scalar inputs.
+ * @return var value with inner Eigen type with result of applying functor to
+ * inputs.
+ *
+ * Note: The return expresssion needs to be evaluated, otherwise the captured
+ *         function and scalar fall out of scope.
+ */
+template <typename T1, typename T2, typename F,
+          require_stan_scalar_t<T1>* = nullptr,
+          require_var_matrix_t<T2>* = nullptr>
+inline auto apply_scalar_binary(const T1& x, const T2& y, const F& f) {
+  return f(x, y);
+}
+
+/**
+ * Specialisation for use when the first input is a nested std::vector and the
+ * second is a scalar. The returned scalar type is deduced to allow for cases
+ * where the input and return scalar types differ (e.g., functions implicitly
+ * promoting integers).
+ *
+ * @tparam T1 Type of std::vector to which functor is applied.
+ * @tparam T2 Type of scalar to which functor is applied.
+ * @tparam F Type of functor to apply.
+ * @param x std::vector input to which operation is applied.
+ * @param y Scalar input to which operation is applied.
+ * @param f functor to apply to inputs.
+ * @return std::vector with result of applying functor to inputs.
+ */
+template <typename T1, typename T2, typename F,
```

Comment:
If you add `is_var_matrix` to the disjunction in `is_container`, then `std::vector<varmat>` will get picked up by the existing overloads in `prim`, and you don't need these.

Reply:
This is a weird thing, since we have like `beta(varmat, varmat)` we actually want to directly call beta. I think I tried this before and it didn't work for some reason. It's a seperate issue, but once I have a bunch of these `var<Matrix>` functions done it should be v easy to add versions that also work for `Matrix<var>` (and Matrix<double>)`. Then imo it could be a good idea to limit the functions that use `apply_scalar_*` to only work for std vectors and std vectors with nested containers.

Right now we have some weird meta programs like 

```template <
    typename T, require_not_var_matrix_t<T>* = nullptr,
    require_all_not_nonscalar_prim_or_rev_kernel_expression_t<T>* = nullptr>
inline auto atanh(const T& x) {
  return apply_scalar_unary<atanh_fun, T>::apply(x);
}
```

and I'd like those to just be 

```cpp
template <typename T, require_container_or_scalar_t<T>* = nullptr>
inline auto atanh(const T& x) {
  return apply_scalar_unary<atanh_fun, T>::apply(x);
}
```

Or something of the sort